sql: allow imprecise pgwire type hints

jordanlewis opened the pull request:
This commit permits the use of imprecise pgwire or sql type hints by
    recording any mismatch between the client-provided type hint and the
    type-system inferred type, and inserting a runtime cast from the
    placeholder's original type to its inferred type.

    Before this patch, type hints that were sent via pgwire or SQL (e.g. in
    `PREPARE x(int) as ...` the `int` is the type hint) were taken perfectly
    literally. Placeholder type hints were required to match the type
    system's inferred type for that placeholder. This simple behavior
    unfortunately leaves us with compatibility issues with some drivers,
    such as JDBC, which send "imprecise type hints" - type hints that don't
    exactly match the correct type for the placeholder.

    The main motivating example is the following query:

    ```
    PREPARE x(INT) AS SELECT * FROM pg_type WHERE oid=$1
    ```

    In this query, the type hint is `INT` but the inferred type is `OID`,
    since we don't support heterogeneous comparisons between integer and
    oid. Previously, this would not type check - but JDBC expects it to.

    After this patch, if the type hint does not match the inferred type, we
    insert a runtime cast from the original type to the inferred type. This
    produces the query `SELECT * FROM pg_type WHERE oid=$1::oid` for our
    example.

    We accomplish this by separately storing the type hints from the
    inferred types. Now, type hints inform the typing of the placeholder
    during prepare, if present. As soon as a placeholder is typed, the final
    placeholder type is set and cannot be changed. If the type hint was
    missing for that placeholder, it is set to the final placeholder type -
    this is required by pgwire so that the values passed to execute can be
    parsed properly.


Previously #14387.
Resolves #14311.
Resolves #14245.
Resolves #14554.
Resolves #17140.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16672)
<!-- Reviewable:end -->

jordanlewis committed:
sql/parser: don't replace placeholders in prepare

And in general when there's no placeholder information in the semantic
context, we don't need to walk the expression tree replacing
placeholders.

jordanlewis committed:
sql/parser: fail on placeholder type mismatch

jordanlewis committed:
sql: permit imprecise type hints

This commit permits the use of imprecise pgwire or sql type hints by
recording any mismatch between the client-provided type hint and the
type-system inferred type, and inserting a runtime cast from the
placeholder's original type to its inferred type.

Before this patch, type hints that were sent via pgwire or SQL (e.g. in
`PREPARE x(int) as ...` the `int` is the type hint) were taken perfectly
literally. Placeholder type hints were required to match the type
system's inferred type for that placeholder. This simple behavior
unfortunately leaves us with compatibility issues with some drivers,
such as JDBC, which send "imprecise type hints" - type hints that don't
exactly match the correct type for the placeholder.

The main motivating example is the following query:

```
PREPARE x(INT) AS SELECT * FROM pg_type WHERE oid=$1
```

In this query, the type hint is `INT` but the inferred type is `OID`,
since we don't support heterogeneous comparisons between integer and
oid. Previously, this would not type check - but JDBC expects it to.

After this patch, if the type hint does not match the inferred type, we
insert a runtime cast from the original type to the inferred type. This
produces the query `SELECT * FROM pg_type WHERE oid=$1::oid` for our
example.

We accomplish this by separately storing the type hints from the
inferred types. Now, type hints inform the typing of the placeholder
during prepare, if present. As soon as a placeholder is typed, the final
placeholder type is set and cannot be changed. If the type hint was
missing for that placeholder, it is set to the final placeholder type -
this is required by pgwire so that the values passed to execute can be
parsed properly.

jordanlewis commented:
Completely reworked. PTAL.

justinj commented:
🎉🎉🎉

---

Review status: 0 of 7 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/acceptance/java_test.go, line 206 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsFuYhFQ11RM9820rI9:-KsFuYhFQ11RM9820rIA:b-mx52jm) ([raw file](https://github.com/cockroachdb/cockroach/blob/af24d2a34b32bfa3664d160ffff2cfdfe56dc2fc/pkg/acceptance/java_test.go#L206)):*
> ```Go
> 		if (!"pg_catalog".equals(nspName)) {
> 		    throw new Exception("unexpected: SELECT can't find pg_catalog database: found " + nspName);
> 		}
> ```

Could you also try selecting `ARRAY[1,2,3]` or something similar? I was running into problems with that before.

---

*[pkg/sql/prepare.go, line 45 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsFuep-_8FNbY_oqXzl:-KsFuep-_8FNbY_oqXzm:b-8xnr2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/af24d2a34b32bfa3664d160ffff2cfdfe56dc2fc/pkg/sql/prepare.go#L45)):*
> ```Go
> 	// dictates how input parameters for those placeholders will be parsed.
> 	SQLTypes       parser.PlaceholderTypes
> 	// OverridenTypes contains the inferred types of the placeholders after
> ```

`Overridden`

---

*[pkg/sql/parser/placeholders.go, line 35 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsFuyf7duqWlaO3mZLy:-KsFuyf7duqWlaO3mZLz:byig381) ([raw file](https://github.com/cockroachdb/cockroach/blob/af24d2a34b32bfa3664d160ffff2cfdfe56dc2fc/pkg/sql/parser/placeholders.go#L35)):*
> ```Go
> 	Values QueryArguments
> 	Types  PlaceholderTypes
> 	// OverriddenTypes relates placeholder names to their overriden type, if any.
> ```

`overridden`

---

*[pkg/sql/parser/placeholders.go, line 139 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsFvK3KTrcphZPHbVhJ:-KsFvK3KTrcphZPHbVhK:byig381) ([raw file](https://github.com/cockroachdb/cockroach/blob/af24d2a34b32bfa3664d160ffff2cfdfe56dc2fc/pkg/sql/parser/placeholders.go#L139)):*
> ```Go
> // case of a placeholder whose type hint does not match the desired type during
> // typechecking. It will return an error if the placeholder never received a
> // type or if the placeholder's type was already overriden.
> ```

`overridden`

---

*[pkg/sql/parser/placeholders.go, line 142 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsFvbqRod7GvcnBYfBh:-KsFvbqRod7GvcnBYfBi:brsgxwy) ([raw file](https://github.com/cockroachdb/cockroach/blob/af24d2a34b32bfa3664d160ffff2cfdfe56dc2fc/pkg/sql/parser/placeholders.go#L142)):*
> ```Go
> func (p *PlaceholderInfo) OverrideType(name string, typ Type) error {
> 	if _, ok := p.Types[name]; !ok {
> 		return fmt.Errorf("cannot override untyped placeholder %s", name)
> ```

Can you use a pgerror here?

---

*[pkg/sql/parser/placeholders.go, line 147 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsFvdfPIqf5ROw0RBqW:-KsFvdfPIqf5ROw0RBqX:bm1ehj8) ([raw file](https://github.com/cockroachdb/cockroach/blob/af24d2a34b32bfa3664d160ffff2cfdfe56dc2fc/pkg/sql/parser/placeholders.go#L147)):*
> ```Go
> 		p.OverriddenTypes = make(map[string]Type, 1)
> 	} else if _, ok := p.OverriddenTypes[name]; ok {
> 		return fmt.Errorf("placeholder %s already has type %s, cannot assign %s", name, p.OverriddenTypes[name], typ)
> ```

also here

---

*[pkg/sql/parser/placeholders.go, line 224 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsG64iK-iP5fBAkH2D5:-KsG64iK-iP5fBAkH2D6:bevl4i1) ([raw file](https://github.com/cockroachdb/cockroach/blob/af24d2a34b32bfa3664d160ffff2cfdfe56dc2fc/pkg/sql/parser/placeholders.go#L224)):*
> ```Go
> // this a no-op and is supported for tests only.
> func replacePlaceholders(expr Expr, ctx *SemaContext) Expr {
> 	if ctx == nil || len(ctx.Placeholders.Values) == 0 {
> ```

just to confirm my understanding - this change is purely an optimization, correct?

---

*[pkg/sql/parser/type_check.go, line 812 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsG5es8mThz0xXNyMx6:-KsG5es8mThz0xXNyMx7:bo4996l) ([raw file](https://github.com/cockroachdb/cockroach/blob/af24d2a34b32bfa3664d160ffff2cfdfe56dc2fc/pkg/sql/parser/type_check.go#L812)):*
> ```Go
> 			//
> 			// Note that we can only override a type once - otherwise we'll permit
> 			// non-well-typed queries like `SELECT NOT $1, $1 > 10` which tries to
> ```

do we have a test of this case?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-:-KsG6VXZjokxh0HMG_ES:b-rorkes)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Review status: 0 of 9 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/sql/prepare.go, line 45 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsFuep-_8FNbY_oqXzl:-KsGzldSfLecER9qlbL0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af24d2a34b32bfa3664d160ffff2cfdfe56dc2fc/pkg/sql/prepare.go#L45)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

`Overridden`
</blockquote></details>

Done.

---

*[pkg/sql/parser/placeholders.go, line 35 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsFuyf7duqWlaO3mZLy:-KsGzn6EmXw5TdkOCt4x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af24d2a34b32bfa3664d160ffff2cfdfe56dc2fc/pkg/sql/parser/placeholders.go#L35)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

`overridden`
</blockquote></details>

Done.

---

*[pkg/sql/parser/placeholders.go, line 139 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsFvK3KTrcphZPHbVhJ:-KsGznWd91ppZCJ9jeUh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af24d2a34b32bfa3664d160ffff2cfdfe56dc2fc/pkg/sql/parser/placeholders.go#L139)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

`overridden`
</blockquote></details>

Done.

---

*[pkg/sql/parser/placeholders.go, line 142 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsFvbqRod7GvcnBYfBh:-KsH3UtG6V3MfMTjXPui:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af24d2a34b32bfa3664d160ffff2cfdfe56dc2fc/pkg/sql/parser/placeholders.go#L142)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Can you use a pgerror here?
</blockquote></details>

Done.

---

*[pkg/sql/parser/placeholders.go, line 147 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsFvdfPIqf5ROw0RBqW:-KsH3VKASeW6JWN8E-UR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af24d2a34b32bfa3664d160ffff2cfdfe56dc2fc/pkg/sql/parser/placeholders.go#L147)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

also here
</blockquote></details>

Done.

---

*[pkg/sql/parser/placeholders.go, line 224 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsG64iK-iP5fBAkH2D5:-KsGzpYglVZ1Vtcewjqh:bb77ufg) ([raw file](https://github.com/cockroachdb/cockroach/blob/af24d2a34b32bfa3664d160ffff2cfdfe56dc2fc/pkg/sql/parser/placeholders.go#L224)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

just to confirm my understanding - this change is purely an optimization, correct?
</blockquote></details>

Yes.

---

*[pkg/sql/parser/type_check.go, line 812 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsG5es8mThz0xXNyMx6:-KsGzsRTLbLIPpNP_Oso:be0b0mh) ([raw file](https://github.com/cockroachdb/cockroach/blob/af24d2a34b32bfa3664d160ffff2cfdfe56dc2fc/pkg/sql/parser/type_check.go#L812)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

do we have a test of this case?
</blockquote></details>

Yes, in TestPGPrepareFail. It actually failed in CI because I didn't account for the case where no pgwire type hint is set.

While fixing this, I realized that there's a better way to express the "overriding" that's happening in this PR. Instead of having types and overridden types, we now have type hints and types in both the placeholder info struct and the prepared statement struct.

Now it's simple to understand what's going on. Type hints inform the first typing of a placeholder. If there's no type hint for a placeholder, it's set by the first typing of the placeholder. As soon as a placeholder is actually typed, its placeholder type is set. That can only happen once. The end.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672)*
<!-- Sent from Reviewable.io -->

justinj commented:
LGTM modulo question!

---

Review status: 0 of 9 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/sql/parser/placeholders.go, line 98 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsJUEebh_U4SX0nxv_B:-KsJUEebh_U4SX0nxv_C:b-f48xxe) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a6caf667020ea7004c3a03a9fc1fa3846c57349/pkg/sql/parser/placeholders.go#L98)):*
> ```Go
> 		return t, true
> 	} else if t, ok := p.TypeHints[name]; ok {
> 		return t, true
> ```

so I don't 100% know what contexts this function will be used in, but is it conceivable that once we've reported a type in this way, we have to "lock in" to it so that if someone attempts to type it differently later on we won't return a different value?

---

*[pkg/sql/parser/type_check.go, line 812 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsG5es8mThz0xXNyMx6:-KsJW-JoR5Lz_6wvCtkQ:b-8eyz07) ([raw file](https://github.com/cockroachdb/cockroach/blob/af24d2a34b32bfa3664d160ffff2cfdfe56dc2fc/pkg/sql/parser/type_check.go#L812)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Yes, in TestPGPrepareFail. It actually failed in CI because I didn't account for the case where no pgwire type hint is set.

While fixing this, I realized that there's a better way to express the "overriding" that's happening in this PR. Instead of having types and overridden types, we now have type hints and types in both the placeholder info struct and the prepared statement struct.

Now it's simple to understand what's going on. Type hints inform the first typing of a placeholder. If there's no type hint for a placeholder, it's set by the first typing of the placeholder. As soon as a placeholder is actually typed, its placeholder type is set. That can only happen once. The end.
</blockquote></details>

Yes this is much more understandable! :thumbsup:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-:-KsJXNkbOwshQap-qULq:bv39wt4)*
<!-- Sent from Reviewable.io -->

justinj commented:
Review status: 0 of 9 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/java_test.go, line 206 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsFuYhFQ11RM9820rI9:-KsJiNKGgzSG1bgPeFsq:b-wb2t6c) ([raw file](https://github.com/cockroachdb/cockroach/blob/af24d2a34b32bfa3664d160ffff2cfdfe56dc2fc/pkg/acceptance/java_test.go#L206)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I'm not sure what you mean here can you give a concrete example?
</blockquote></details>

I just mean try running a query like `SELECT ARRAY[1,2,3]` and reading it, when I tried this before I would get a type error from some system query jdbc would issue when running it, for some reason.

---

*[pkg/sql/parser/placeholders.go, line 98 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsJUEebh_U4SX0nxv_B:-KsJigsARBgrpNTX6xbz:b5o90re) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a6caf667020ea7004c3a03a9fc1fa3846c57349/pkg/sql/parser/placeholders.go#L98)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

We do have to do that, but unfortunately it needs to be up to the caller, since at the time that we get a type hint we might choose to lock in to a different type. This is done in Placeholder.TypeCheck.
</blockquote></details>

I see, ok!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672)*
<!-- Sent from Reviewable.io -->

knz commented:
I think this is OK, I have a few minor comments below.
However before this merges I would like you to extend the commit message and make a longer clarifying story about what is going on here, as an overview. The story should contain:

- what was happening before this patch
- what is happening now
- how you made it happen

For the two first points use an example query and show separately for prepare/execute what happens with their placeholders.

---

Reviewed 7 of 7 files at r4, 2 of 2 files at r5, 9 of 9 files at r6.
Review status: all files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[pkg/sql/logictest/testdata/logic_test/prepare, line 251 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsJjX9czw1QFukQNi1H:-KsJjX9dgFSuezN-AlRB:btvc37a) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a6caf667020ea7004c3a03a9fc1fa3846c57349/pkg/sql/logictest/testdata/logic_test/prepare#L251)):*
> ```
> # Mismatch between expected and hinted types should prepare, but potentially
> # fail to execute if the cast is not possible.
> statement
> ```

nit: say "statement ok" here and below

---

*[pkg/sql/parser/placeholders.go, line 224 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsG64iK-iP5fBAkH2D5:-KsJh_MkfdsvTKamFvrK:b-rpg1ew) ([raw file](https://github.com/cockroachdb/cockroach/blob/af24d2a34b32bfa3664d160ffff2cfdfe56dc2fc/pkg/sql/parser/placeholders.go#L224)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Yes.
</blockquote></details>

This deserves an additional comment: your optimization and, "in the case placeholders are present in the SQL but none are provided during execute, the further type analysis will detect this and report an unassigned placeholder error". (This case is checked in a test, right?)

---

*[pkg/sql/parser/placeholders.go, line 34 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsJjxBhi2wAsAft1Qqo:-KsJjxBiM_BC39G_vIVL:btyd26q) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a6caf667020ea7004c3a03a9fc1fa3846c57349/pkg/sql/parser/placeholders.go#L34)):*
> ```Go
> // PlaceholderInfo defines the interface to SQL placeholders.
> type PlaceholderInfo struct {
> 	Values    QueryArguments
> ```

Please document the difference between the two fields here.

---

*[pkg/sql/parser/placeholders.go, line 98 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsJUEebh_U4SX0nxv_B:-KsJlLaLYdMaunVXcacD:biognct) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a6caf667020ea7004c3a03a9fc1fa3846c57349/pkg/sql/parser/placeholders.go#L98)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I see, ok!
</blockquote></details>

It seems to me that (*PlaceholderInfo).Type() can never be called until Placeholder.TypeCheck has been closed. So it's not so clear to me why this case needs to be handled here. For the use of the hint made in Placeholder.TypeCheck, we could have a separate method (PlaceholderInfo.TypeHint?)

---

*[pkg/sql/parser/placeholders.go, line 185 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsJmAosu9J0taaMHNUf:-KsJmAotBjYxve0A0PG1:b4e67y8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a6caf667020ea7004c3a03a9fc1fa3846c57349/pkg/sql/parser/placeholders.go#L185)):*
> ```Go
> 			if !typed {
> 				// All placeholders should be typed at this point. If they're not, we
> 				// can just return their value, though.
> ```

Ugh, why not an error?

---

*[pkg/sql/parser/placeholders.go, line 189 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsJmGGKRQA6WMi4LDQf:-KsJmGGLIVSyNV02bwM6:b-q2m3an) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a6caf667020ea7004c3a03a9fc1fa3846c57349/pkg/sql/parser/placeholders.go#L189)):*
> ```Go
> 			}
> 			if !e.ResolvedType().Equivalent(typ) {
> 				// This happens when we overrode the placeholder's type during type
> ```

override

---

*[pkg/sql/parser/placeholders.go, line 195 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsJmKJ3WvXq0Zvm4Chs:-KsJmKJ3WvXq0Zvm4Cht:bovgh2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a6caf667020ea7004c3a03a9fc1fa3846c57349/pkg/sql/parser/placeholders.go#L195)):*
> ```Go
> 				// TODO(jordan): introduce a restriction on what casts are allowed here.
> 				colType, err := DatumTypeToColumnType(typ)
> 				if err != nil {
> ```

I don't think the error should be silently dropped.

---

*[pkg/sql/parser/type_check.go, line 814 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsJmWsQEBg9b_f241SF:-KsJmWsR9U8DTcqshDJq:bxfu4db) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a6caf667020ea7004c3a03a9fc1fa3846c57349/pkg/sql/parser/type_check.go#L814)):*
> ```Go
> 			// the type system expects. Then, when the value is actually sent to us
> 			// later, we cast the input value (whose type is the expected type) to the
> 			// desired type here.
> ```

There word "here" in this sentence is unclear. Where is the cast happening?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-:-KsJmmQ8qNKqJvXbRmRm:b-dtvn1v)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/java_test.go, line 206 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsFuYhFQ11RM9820rI9:-KsJnBPE487DZ5nmR4hU:bxv4l7a) ([raw file](https://github.com/cockroachdb/cockroach/blob/af24d2a34b32bfa3664d160ffff2cfdfe56dc2fc/pkg/acceptance/java_test.go#L206)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I just mean try running a query like `SELECT ARRAY[1,2,3]` and reading it, when I tried this before I would get a type error from some system query jdbc would issue when running it, for some reason.
</blockquote></details>

Hmmm but that has nothing to do with this patch because there's no placeholders there. I don't expect it to change.

I did a test with a placeholder and it has a different problem that is not related to this patch:

```
//CREATE TABLE test.v (x int[])
stmt = conn.prepareStatement("INSERT INTO test.v VALUES(?)");
Array array = conn.createArrayOf("INT8", new Object[]{1,2,3});
```

This produces: Exception in thread "main" org.postgresql.util.PSQLException: Unable to find server array type for provided name INT8.

We should open an issue for this but it's not related to this patch.

---

*[pkg/sql/logictest/testdata/logic_test/prepare, line 251 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsJjX9czw1QFukQNi1H:-KsJqGcUqRsT73-a2ut6:b3unalr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a6caf667020ea7004c3a03a9fc1fa3846c57349/pkg/sql/logictest/testdata/logic_test/prepare#L251)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: say "statement ok" here and below
</blockquote></details>

This whole file is missing `ok`. I will fix in a separate pr.

---

*[pkg/sql/parser/placeholders.go, line 224 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsG64iK-iP5fBAkH2D5:-KsJxPxY35RHm7MZf24f:b-85zd28) ([raw file](https://github.com/cockroachdb/cockroach/blob/af24d2a34b32bfa3664d160ffff2cfdfe56dc2fc/pkg/sql/parser/placeholders.go#L224)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This deserves an additional comment: your optimization and, "in the case placeholders are present in the SQL but none are provided during execute, the further type analysis will detect this and report an unassigned placeholder error". (This case is checked in a test, right?)
</blockquote></details>

Done. Detecting missing placeholders is done before this point for both pgwire execute and sql execute.

---

*[pkg/sql/parser/placeholders.go, line 34 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsJjxBhi2wAsAft1Qqo:-KsJqMSGHFYSVSjJM7F9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a6caf667020ea7004c3a03a9fc1fa3846c57349/pkg/sql/parser/placeholders.go#L34)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please document the difference between the two fields here.
</blockquote></details>

Done.

---

*[pkg/sql/parser/placeholders.go, line 98 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsJUEebh_U4SX0nxv_B:-KsJwQNEAggy6YVhQNWQ:b-177u3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a6caf667020ea7004c3a03a9fc1fa3846c57349/pkg/sql/parser/placeholders.go#L98)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

It seems to me that (*PlaceholderInfo).Type() can never be called until Placeholder.TypeCheck has been closed. So it's not so clear to me why this case needs to be handled here. For the use of the hint made in Placeholder.TypeCheck, we could have a separate method (PlaceholderInfo.TypeHint?)
</blockquote></details>

It's called in two places, during type checking and then later. I think it's more confusing than helpful to segregate the methods, since during type checking we do have to examine both.

---

*[pkg/sql/parser/placeholders.go, line 185 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsJmAosu9J0taaMHNUf:-KsJwfFCImdl1RVa4bdU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a6caf667020ea7004c3a03a9fc1fa3846c57349/pkg/sql/parser/placeholders.go#L185)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Ugh, why not an error?
</blockquote></details>

Done.

---

*[pkg/sql/parser/placeholders.go, line 189 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsJmGGKRQA6WMi4LDQf:-KsJwfisU9ii3KLeBYCR:bj262oa) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a6caf667020ea7004c3a03a9fc1fa3846c57349/pkg/sql/parser/placeholders.go#L189)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

override
</blockquote></details>

No, I'm using the past tense here on purpose.

---

*[pkg/sql/parser/placeholders.go, line 195 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsJmKJ3WvXq0Zvm4Chs:-KsJwnkAQyoYp7aKjpKs:b7sn07w) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a6caf667020ea7004c3a03a9fc1fa3846c57349/pkg/sql/parser/placeholders.go#L195)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I don't think the error should be silently dropped.
</blockquote></details>

You're right, Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
TFTRs!

knz commented:
Great commit message :100:

There's still an error message to fix so you're up for another round of CI anyway, so let's finish the discussion on placeholderinfo.Type().

---

Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/placeholders.go, line 98 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsJUEebh_U4SX0nxv_B:-KsKMAFZTUA5NjoSQ_Zs:b-g0r3le) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a6caf667020ea7004c3a03a9fc1fa3846c57349/pkg/sql/parser/placeholders.go#L98)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

It's called in two places, during type checking and then later. I think it's more confusing than helpful to segregate the methods, since during type checking we do have to examine both.
</blockquote></details>

But that's the thing right? After type checking, we should *never* be looking at the hints map. Right now if there's a bug and we end up not populating p.Types properly, this method will silently mask that and make PlaceholderInfo.Type respond something because there's still a hint laying around. That's why I would like to find a way to assert the hints are only reachable during type checking and not afterwards.

---

*[pkg/sql/parser/placeholders.go, line 193 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsKMUUj1UKxvzSyi-X5:-KsKMUUj1UKxvzSyi-X6:b-n0lgac) ([raw file](https://github.com/cockroachdb/cockroach/blob/04f80b1881cfbefb296bc21795800c813fecb940/pkg/sql/parser/placeholders.go#L193)):*
> ```Go
> 			if !typed {
> 				// All placeholders should be typed at this point.
> 				v.err = pgerror.NewErrorf(pgerror.CodeInternalError, "Missing type for placeholder %s", t.Name)
> ```

lowercase

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-:-KsKMb9noKmERMRRrUOF:b-l4qrev)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/placeholders.go, line 98 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsJUEebh_U4SX0nxv_B:-KsKq1IgSt3gHx0V59Na:b-eqhzau) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a6caf667020ea7004c3a03a9fc1fa3846c57349/pkg/sql/parser/placeholders.go#L98)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

But that's the thing right? After type checking, we should *never* be looking at the hints map. Right now if there's a bug and we end up not populating p.Types properly, this method will silently mask that and make PlaceholderInfo.Type respond something because there's still a hint laying around. That's why I would like to find a way to assert the hints are only reachable during type checking and not afterwards.
</blockquote></details>

Okay, I added a boolean flag to the function `allowTypeHints`.

---

*[pkg/sql/parser/placeholders.go, line 193 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672#-KsKMUUj1UKxvzSyi-X5:-KsKl6dq3PzBqR8oyxBi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/04f80b1881cfbefb296bc21795800c813fecb940/pkg/sql/parser/placeholders.go#L193)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

lowercase
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16672)*
<!-- Sent from Reviewable.io -->

knz commented:
LGTM
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

jordanlewis commented:
Woohoo! TFTRs! This one started in March - psyched to get it in.
